go-controller/pkg/ovn/ipallocator/allocator.go

```diff
@@ -201,49 +216,8 @@ func (r *Range) contains(ip net.IP) (bool, int) {
 	return true, offset
 }
 
-// bigForIP creates a big.Int based on the provided net.IP
-func bigForIP(ip net.IP) *big.Int {
-	b := ip.To4()
-	if b == nil {
-		b = ip.To16()
-	}
-	return big.NewInt(0).SetBytes(b)
-}
-
-// addIPOffset adds the provided integer offset to a base big.Int representing a
-// net.IP
-func addIPOffset(base *big.Int, offset int) net.IP {
-	return net.IP(big.NewInt(0).Add(base, big.NewInt(int64(offset))).Bytes())
-}
-
 // calculateIPOffset calculates the integer offset of ip from base such that
 // base + offset = ip. It requires ip >= base.
 func calculateIPOffset(base *big.Int, ip net.IP) int {
-	return int(big.NewInt(0).Sub(bigForIP(ip), base).Int64())
-}
-
-// RangeSize returns the size of a range in valid addresses.
-func RangeSize(subnet *net.IPNet) int64 {
-	ones, bits := subnet.Mask.Size()
-	if bits == 32 && (bits-ones) >= 31 || bits == 128 && (bits-ones) >= 127 {
-		return 0
-	}
-	// For IPv6, the max size will be limited to 65536
-	// This is due to the allocator keeping track of all the
-	// allocated IP's in a bitmap. This will keep the size of
-	// the bitmap to 64k.
-	if bits == 128 && (bits-ones) >= 16 {
-		return int64(1) << uint(16)
-	} else {
-		return int64(1) << uint(bits-ones)
-	}
-}
-
-// GetIndexedIP returns a net.IP that is subnet.IP + index in the contiguous IP space.
-func GetIndexedIP(subnet *net.IPNet, index int) (net.IP, error) {
-	ip := addIPOffset(bigForIP(subnet.IP), index)
-	if !subnet.Contains(ip) {
-		return nil, fmt.Errorf("can't generate IP with index %d from subnet. subnet too small. subnet: %q", index, subnet)
-	}
-	return ip, nil
+	return int(big.NewInt(0).Sub(utilnet.BigForIP(ip), base).Int64())
```

Comment:
actually I missed this on initial pass, i think this will still overflow let me check

Reply:
yeah this will all still overflow, you need to fix that and add test cases:
https://play.golang.org/p/VBvpkW4NCJd

Reply:
Is it tho? I can see how this private function is not bound, but in the constructor NewAllocatorCIDRRange we call the RangeSize function which now contains the fix from upstream for not overflowing:

https://github.com/kubernetes/utils/commit/50ce1e46fdafcb238e4b0a9c88eae5d5238f68a2

I guess what I'm trying to say, if we want to add a test to calculeIPOffset I don't see that test in upstream, so if applicable push upstream, then bring it down.

@aojea @danwinship since you have stuff in that part of the code upstream, thoughts?

Reply:
you can get `GetIndexedIP` from utilnet too, all these functions are exported

Reply:
sorry, the thing is that the code upstream checks that the IP belongs to the allocator range, this way it never overflows

Reply:
and this does the same 
```
// contains returns true and the offset if the ip is in the range, and false
// and nil otherwise. The first and last addresses of the CIDR are omitted.
func (r *Range) contains(ip net.IP) (bool, int) {
	if !r.net.Contains(ip) {
		return false, 0
	}
```

Reply:
and here too
```
	ip := addIPOffset(bigForIP(subnet.IP), index)
	if !subnet.Contains(ip) {
		return nil, fmt.Errorf("can't generate IP with index %d from subnet. subnet too small. subnet: %q", index, subnet)
	}
```

Reply:
so the logic is that you define a range that is bounded, and then you choose IP inside that range, that guarantees you to never overflow, because you are working in a bounded dimension ...

Reply:
exactly,

Reply:
I don't think so guys... "Has" is an exported function:
```
// Has returns true if the provided IP is already allocated and a call
// to Allocate(ip) would fail with ErrAllocated.
func (r *Range) Has(ip net.IP) bool {
	ok, offset := r.contains(ip)
	if !ok {
		return false
	}

	return r.alloc.Has(offset)
}
```

Which calls contains, which will run:
```
// contains returns true and the offset if the ip is in the range, and false
// and nil otherwise. The first and last addresses of the CIDR are omitted.
func (r *Range) contains(ip net.IP) (bool, int) {
	if !r.net.Contains(ip) {
		return false, 0
	}

	offset := calculateIPOffset(r.base, ip)
	if offset < 0 || offset >= r.max {
		return false, 0
	}
	return true, offset
}
```

r.net is set to the cidr given, so of course it will contain the IP given in my goplayground example. That means we will then call calculateIPOffset, which will overflow to a negative number, and then return false, when really it is true.

Reply:
if IP is within the range, (IP-Base) can't be never > int64

Reply:
hmm, IC the problem this code deviates from upstream, in upstream you can't allocate that is out of the range.
The range here is capped to 65536. so the contains() will not allow something that  (IP-Base) > 65536

Reply:
in upstream the CIDR are limited to /112 for IPv6 and to /12 for IPv6, you can not set a /64, this is something I'm trying to change for more than one year but is not straightforward
https://github.com/kubernetes/enhancements/pull/1881

Reply:
Yeah, that seems right now:

https://play.golang.org/p/ZHz9DYZdGro

I flattened in a playground the patch code, and if you run a Has, it doesn't return true, which I guess is due to /64 range, which is not a thing in upstream and thus why they don't hit.